File: apps/client-macos/Sources/TimeTrack/Projects/ProjectClient.swift
import Foundation

/// Fetches the team's projects (with tasks) for the picker. GET /v1/projects is available
/// to any authenticated user and is team-scoped server-side. On a 401 it forces a token
/// refresh and retries once (mirrors PolicyClient); any other failure propagates so the
/// caller falls back to the cache. `includeArchived` is omitted → the API defaults to
/// assignable-only projects.
enum ProjectClientError: Error { case unavailable }

final class ProjectClient {
    private let baseURL: URL
    private let session: AuthSession

    init(baseURL: URL, session: AuthSession) {
        self.baseURL = baseURL
        self.session = session
    }

    func list() async throws -> [Project] {
        let token = try await session.accessToken()
        let (data, status) = try await fetch(token: token)
        if status == 401 {
            let refreshed = try await session.forceRefresh()
            let (data2, status2) = try await fetch(token: refreshed)
            guard status2 == 200 else { throw ProjectClientError.unavailable }
            return try JSONDecoder().decode([Project].self, from: data2)
        }
        guard status == 200 else { throw ProjectClientError.unavailable }
        return try JSONDecoder().decode([Project].self, from: data)
    }

    private func fetch(token: String) async throws -> (Data, Int) {
        var request = URLRequest(url: baseURL.appendingPathComponent("projects"))
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, response) = try await URLSession.shared.data(for: request)
        return (data, (response as? HTTPURLResponse)?.statusCode ?? 0)
    }
}

File: apps/client-windows/src/NiftyTimer/Projects/ProjectClient.cs
using NiftyTimer.Auth;

namespace NiftyTimer.Projects;

public interface IProjectClient
{
    Task<IReadOnlyList<Project>> ListAsync(CancellationToken cancellationToken = default);
}

/// <summary>
/// Fetches the team's projects (with tasks) for the picker. <c>GET /v1/projects</c> is available
/// to any authenticated user and is team-scoped server-side — it deliberately carries no
/// <c>@Roles</c> precisely so the desktop client can call it. <c>includeArchived</c> is omitted,
/// so the API returns assignable-only projects.
///
/// Any failure propagates so the caller falls back to <see cref="ProjectCache"/> rather than
/// showing an empty picker.
/// </summary>
public sealed class ProjectClient : IProjectClient
{
    private readonly AuthorizedJsonClient _json;

    public ProjectClient(AuthorizedJsonClient json) => _json = json;

    public async Task<IReadOnlyList<Project>> ListAsync(CancellationToken cancellationToken = default) =>
        await _json.GetAsync<List<Project>>("projects", cancellationToken).ConfigureAwait(false);
}
